source/dem/insertion.cc

```diff
@@ -197,5 +221,131 @@ Insertion<dim>::calculate_insertion_domain_maximum_particle_number(
+          // We need to add the cell iterator to the edge_of_box container.
+
+          // If the cell is outside, both bool will be false, and we do nothing.
+          if (completely_inside)
+            {
+              in_the_clearing_box.insert(cell);
+            }
+          else if (!completely_inside && partially_inside)
+            {
+              edge_of_box.insert(cell);
+            }
+        }
+    }
+}
+
+template <int dim>
+void
+Insertion<dim>::clear_particle_box(
+  Particles::ParticleHandler<dim> &particle_handler)
+{
+  std::vector<
+    typename dealii::Particles::ParticleHandler<dim>::particle_iterator>
+    to_remove_iterators;
+
+  // Reserve to the maximum number of particle on this proc
+  to_remove_iterators.reserve(particle_handler.n_locally_owned_particles());
+
+  // Loop over the first container
+  for (const auto &cell_in_box : in_the_clearing_box)
+    {
+      // Check if this cell has particles
+      auto particles_in_cell = particle_handler.particles_in_cell(cell_in_box);
+
+      if (!particles_in_cell.empty())
+        {
+          // Loop over the particle in the cell
+          for (auto particle_in_cell = particles_in_cell.begin();
+               particle_in_cell != particles_in_cell.end();
+               ++particle_in_cell)
+            {
+              // Since we know the cell is fully inside the box, we can
+              // delete every particles in it.
+
+              to_remove_iterators.push_back(particle_in_cell);
+            }
+        }
+    }
+
+  // Loop over the second container
+  for (auto cell_edge_of_box = edge_of_box.begin();
+       cell_edge_of_box != edge_of_box.end();
+       ++cell_edge_of_box)
+    {
+      // Check if this cell has particle
+      auto particles_in_cell =
+        particle_handler.particles_in_cell(*cell_edge_of_box);
+      const bool particles_exist_in_cell = !particles_in_cell.empty();
+
+      if (particles_exist_in_cell)
+        {
+          // Loop over the particles in the cell
+          for (auto particle_in_cell = particles_in_cell.begin();
+               particle_in_cell != particles_in_cell.end();
+               ++particle_in_cell)
+            {
+              // We need to check if the particle is inside since the cell
+              // is at the edge.
+              Point<3> particle_position =
+                point_nd_to_3d(particle_in_cell->get_location());
+              if (p_min[0] <= particle_position[0] &&
+                  p_max[0] >= particle_position[0] &&
+                  p_min[1] <= particle_position[1] &&
+                  p_max[1] >= particle_position[1] &&
+                  p_min[2] <= particle_position[2] &&
+                  p_max[2] >= particle_position[2])
+                {
+                  to_remove_iterators.push_back(particle_in_cell);
+                }
+            }
+        }
+    }
+  particle_handler.remove_particles(to_remove_iterators);
```

Comment:
Instead of doing this, why don't you just remove particles one by one if they fulfill the condition. Pushing back will add an unnecessary cost here.

Reply:
It's better to remove all of them. It's expensive to remove one by one because the particle_handler needs to restructure its cache.
In. general, always insert or remove as chunks and not one by one. This is the correct implementation.

Reply:
Good to know. Thanks!